src/commands/login.rs

```diff
@@ -1,29 +1,121 @@
+use anyhow::format_err;
+use opener;
+use serde::Deserialize;
+use std::path::PathBuf;
+use std::{thread, time};
 use structopt::StructOpt;
 
-use crate::auth::AuthStore;
+use crate::{auth::AuthStore, manifest::Manifest, package_index::PackageIndex};
 
 /// Log into a registry.
 #[derive(Debug, StructOpt)]
 pub struct LoginSubcommand {
-    /// Authentication token for the registry. If not specified, Wally will
-    /// prompt.
-    pub token: Option<String>,
+    /// Path to a project to decide how to login
+    #[structopt(long = "project-path", default_value = ".")]
+    pub project_path: PathBuf,
+}
+
+#[derive(Debug, Deserialize)]
+struct DeviceCodeResponse {
+    device_code: String,
+    user_code: String,
+    verification_uri: String,
+    expires_in: u64,
+    interval: u64,
+}
+
+#[derive(Deserialize)]
+struct DeviceCodeAuth {
+    access_token: String,
+    token_type: String,
+    scope: String,
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum AuthResponse {
+    Ok(DeviceCodeAuth),
+    Err { error: String },
+}
+
+fn wait_for_github_auth(
+    device_code_response: DeviceCodeResponse,
+    oauth_id: &str,
+) -> anyhow::Result<DeviceCodeAuth> {
+    let client = reqwest::blocking::Client::new();
+
+    thread::sleep(time::Duration::from_secs(device_code_response.interval));
```

Comment:
Let's import `sleep`, `Duration`, and `Client` explicitly instead of using their fully qualified paths here.